browser/component_updater/brave_component_contents_verifier.cc

```diff
@@ -0,0 +1,171 @@
+constexpr char kBraveVerifiedContentsPath[] =
+    "brave_metadata/verified_contents.json";
+
+std::string GetRootHasheForContent(base::span<const uint8_t> contents,
+                                   size_t block_size) {
+  std::vector<std::string> hashes;
+  // Even when the contents is empty, we want to output at least one hash
+  // block (the hash of the empty string).
+  base::SpanReader reader(contents);
+  do {
+    auto chunk = reader.Read(std::min(block_size, reader.remaining()));
+    std::unique_ptr<crypto::SecureHash> hash(
+        crypto::SecureHash::Create(crypto::SecureHash::SHA256));
+    hash->Update(chunk.value());
+
+    std::string buffer;
+    buffer.resize(crypto::kSHA256Length);
+    hash->Finish(base::as_writable_byte_span(buffer));
+    hashes.push_back(std::move(buffer));
+  } while (reader.remaining() > 0);
+
+  CHECK(block_size % crypto::kSHA256Length == 0);
+  return extensions::ComputeTreeHashRoot(hashes,
+                                         block_size / crypto::kSHA256Length);
+}
+
+// Proxies the file access and checks the verified_contents.json.
+class ComponentContentsAccessorImpl
+    : public brave_component_updater::ComponentContentsAccessor {
+ public:
+  explicit ComponentContentsAccessorImpl(const base::FilePath& component_root)
+      : brave_component_updater::ComponentContentsAccessor(component_root) {
+    if (base::PathExists(
+            component_root.AppendASCII(kBraveVerifiedContentsPath))) {
+      verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+          kComponentContentsVerifierPublicKey,
+          component_root.AppendASCII(kBraveVerifiedContentsPath));
+    } else {
+      verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+          kComponentContentsVerifierPublicKey,
+          component_root.AppendASCII(kVerifiedContentsPath));
+    }
+    if (verified_contents_ &&
+        verified_contents_->block_size() % crypto::kSHA256Length != 0) {
+      // Unsupported block size.
+      verified_contents_.reset();
+    }
+  }
+
+  bool IsComponentSignatureValid() const override {
+    return !!verified_contents_.get();
+  }
+
+  bool VerifyContents(const base::FilePath& relative_path,
+                      base::span<const uint8_t> contents) override {
+    if (!IsComponentSignatureValid()) {
+      return false;
+    }
+
+    if (!verified_contents_->HasTreeHashRoot(relative_path)) {
+      return false;
+    }
+
+    const auto root_hash =
+        GetRootHasheForContent(contents, verified_contents_->block_size());
+    return verified_contents_->TreeHashRootEquals(relative_path, root_hash);
+  }
+
+ private:
+  ~ComponentContentsAccessorImpl() override = default;
+
+  std::unique_ptr<extensions::VerifiedContents> verified_contents_;
+};
+
+bool ShouldBypassSignature() {
+  static const bool kBypass =
+      !base::FeatureList::IsEnabled(
+          component_updater::kComponentContentsVerifier) ||
```

Comment:
Do we suppose to inline the feature after the launch?
Otherwise using bad_flags doesn't make much sense (can be easily bypassed via `--disable-features= ComponentContentsVerifier`).